## CLEANING DATA – nba.csv

**This task is designed to encourage you to:  
•	Use the knowledge you have acquired in your previous tasks and apply it here  
•	Search the internet for solutions to anything you do not know how to do yet  
    Tip: start all internet search criteria with the word Pandas**

In [7]:
#import the relevant libraries - you will need pandas and numpy
import pandas as pd
import numpy as np

In [8]:
#read the csv file and assign it to a dataframe (download the file and ensure that it is in the same folder as this ipynb file)
df = pd.read_csv('nba.csv')

#describe the dataframe
df

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   
457            NaN             NaN     NaN      NaN   NaN    NaN     NaN   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        NaN  
3        Georgia State  1148640.0  
4                  NaN  5000000.0  
..                 ...        ...  
453             Butler  2433333.0  
454                NaN   900000.0  
455                NaN  2900000.0  
456             Kansas   947276.0  
457                NaN        NaN  

[458 rows x 9 columns]

In [9]:
#Display the shape of the data - how many rows and columns  do you have?
df.shape

(458, 9)

In [10]:
#Display the head of the document - this will give you an idea of what the contents of the file looks like
#The first five rows are displayed
df.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [11]:
#Display the last 5 lines of your file
#What problem is immediateley obvious when looking at the last five lines of the file?
df.tail()

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

In [12]:
#Fix the problem by removing the row. Ensure that the dataframe has been changed
# Remove the row with index 457 containing all null values
df.drop(index=457, inplace=True)

# Check that the row has been removed by printing the updated DataFrame
df

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        NaN  
3        Georgia State  1148640.0  
4                  NaN  5000000.0  
..                 ...        ...  
452           Kentucky  2239800.0  
453             Butler  2433333.0  
454                NaN   900000.0  
455                NaN  2900000.0  
456             Kansas   947276.0  

[457 rows x 9 columns]

In [13]:
#Examine the different datatypes - we need to understand the data types as it effects what we can do with the data
df.dtypes

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

In [14]:
#Count the number of values in each column
df.count()

Name        457
Team        457
Number      457
Position    457
Age         457
Height      457
Weight      457
College     373
Salary      446
dtype: int64

In [15]:
#Count the number of records in a specific column - NAme column
#df.count()[0]

#Or 

df.count()['Name']

457

In [16]:
#Count the number of individual values in one column
df["Team"].value_counts()

Team
New Orleans Pelicans      19
Memphis Grizzlies         18
New York Knicks           16
Milwaukee Bucks           16
Boston Celtics            15
Brooklyn Nets             15
Portland Trail Blazers    15
Oklahoma City Thunder     15
Denver Nuggets            15
Washington Wizards        15
Miami Heat                15
Charlotte Hornets         15
Atlanta Hawks             15
San Antonio Spurs         15
Houston Rockets           15
Dallas Mavericks          15
Indiana Pacers            15
Detroit Pistons           15
Cleveland Cavaliers       15
Chicago Bulls             15
Sacramento Kings          15
Phoenix Suns              15
Los Angeles Lakers        15
Los Angeles Clippers      15
Golden State Warriors     15
Toronto Raptors           15
Philadelphia 76ers        15
Utah Jazz                 15
Orlando Magic             14
Minnesota Timberwolves    14
Name: count, dtype: int64

In [17]:
#Find the maximum Salary
#MAX - to calculate this use max()

#read in a fresh copy of your data file to temp_df
temp_df = pd.read_csv('nba.csv')

#calculate the maximum value of the Salary column
x = temp_df['Salary'].max()
#print the maximum value
print("Max:" + str(x))

#Use the maximum value to fill null values

#Use x to fill null values
temp_df['Salary'].fillna(x, inplace = True)

#display temp_df - check rows 18 and 28
temp_df

Max:25000000.0


Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   
457            NaN             NaN     NaN      NaN   NaN    NaN     NaN   

               College      Salary  
0                Texas   7730337.0  
1            Marquette   6796117.0  
2    Boston University  25000000.0  
3        Georgia State   1148640.0  
4                  NaN   5000000.0  
..                 ...         ...  
453             Butler   2433333.0  
454                NaN    900000.0  
455                NaN   2900000.0  
456             Kansas    947276.0  
457                NaN  25000000.0  

[458 rows x 9 columns]

In [18]:
#Find the minimum Weight


In [19]:
#Find maximum Height
#MAX - to calculate this use max()

#read in a fresh copy of your data file to temp_df
temp_df = pd.read_csv('nba.csv')

#calculate the maximum value of the Height column
x = temp_df['Height'].max()
#print the maximum value
print("Max:" + str(x))

#Use the maximum value to fill null values

#Use x to fill null values
#temp_df['Salary'].fillna(x, inplace = True)

#display temp_df - check rows 18 and 28
temp_df

TypeError: '>=' not supported between instances of 'str' and 'float'

In [20]:
#The maximum Height code above actually works even though it is Object data
#Run this code below to see where a potential problem could arise

x = df["Height"].max()
y = df["Height"].min()
z= x-y
z

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [21]:
#Describe the Object data in the dataframe
df.describe(include = "O" )
    #What conclusions can be made from this? NOT ALL WENT TO COLLEGE
    #Which column has the highest number of unique values? NAME
    #Which column has too many missing values? - We couldn't reasonably caluclate these.



Name                  Team Position Height   College
count             457                   457      457    457       373
unique            457                    30        5     18       118
top     Avery Bradley  New Orleans Pelicans       SG    6-9  Kentucky
freq                1                    19      102     59        22

In [22]:
#Delete the column with highest number of unique values & the column with the highest number of missing values & update the dataframe


#Display the dataframe

# Count missing values in each column
missing_count = df.isnull().sum()
#missing_count

# Identify the column with the highest number of missing values
max_missing_col = missing_count.idxmax()
#max_missing_col

# Count the number of unique values in each object column
unique_counts = df.select_dtypes(include='O').nunique()

# Identify the column with the highest number of unique values
max_unique_col = unique_counts.idxmax()

# Display the column name with the highest number of unique values
#max_unique_col
df.drop(columns=max_unique_col, inplace=True)


# Remove both columns
df.drop(columns=max_missing_col, inplace=True)
df
# Display the updated dataframe
#df




Team  Number Position   Age Height  Weight     Salary
0    Boston Celtics     0.0       PG  25.0    6-2   180.0  7730337.0
1    Boston Celtics    99.0       SF  25.0    6-6   235.0  6796117.0
2    Boston Celtics    30.0       SG  27.0    6-5   205.0        NaN
3    Boston Celtics    28.0       SG  22.0    6-5   185.0  1148640.0
4    Boston Celtics     8.0       PF  29.0   6-10   231.0  5000000.0
..              ...     ...      ...   ...    ...     ...        ...
452       Utah Jazz    41.0       PF  20.0   6-10   234.0  2239800.0
453       Utah Jazz     8.0       PG  26.0    6-3   203.0  2433333.0
454       Utah Jazz    25.0       PG  24.0    6-1   179.0   900000.0
455       Utah Jazz    21.0        C  26.0    7-3   256.0  2900000.0
456       Utah Jazz    24.0        C  26.0    7-0   231.0   947276.0

[457 rows x 7 columns]

In [23]:
#Describe the numerical data
df.describe()

Number         Age      Weight        Salary
count  457.000000  457.000000  457.000000  4.460000e+02
mean    17.678337   26.938731  221.522976  4.842684e+06
std     15.966090    4.404016   26.368343  5.229238e+06
min      0.000000   19.000000  161.000000  3.088800e+04
25%      5.000000   24.000000  200.000000  1.044792e+06
50%     13.000000   26.000000  220.000000  2.839073e+06
75%     25.000000   30.000000  240.000000  6.500000e+06
max     99.000000   40.000000  307.000000  2.500000e+07

When we described the data above, the values that were returned for the Salary field don’t look quite right – however they are. It’s just that there is not enough space to display all of the digits.  
* If you look at the results of the Count field in the Salary column above, the number is displayed as 4.460000e+02.  
* To understand the number move the decimal point by 2 place to the right (that's what e+02 means), giving 446  
    The other colums have 457 records

With code, count the number of null values in the dataframe. It will be 11.  
* The number of records at this point is 457. So we can conclude that the count value is actually correct.

In [24]:
#Sum the null values in columns 
# Count the number of null values in each column
null_counts = df.isnull().sum()

# Display the null counts
null_counts


Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
Salary      11
dtype: int64

To address the null values, you have 2 choices at this point
* Calculate the mean in the Salary field and replace all occurrences of NaN with the mean salary  
Or
* Delete the rows with null values

In [25]:
#Display the dataframe - row 2 (3rd record) has a null value in the Salary column
df.head()

Team  Number Position   Age Height  Weight     Salary
0  Boston Celtics     0.0       PG  25.0    6-2   180.0  7730337.0
1  Boston Celtics    99.0       SF  25.0    6-6   235.0  6796117.0
2  Boston Celtics    30.0       SG  27.0    6-5   205.0        NaN
3  Boston Celtics    28.0       SG  22.0    6-5   185.0  1148640.0
4  Boston Celtics     8.0       PF  29.0   6-10   231.0  5000000.0

In [26]:
#Calculate the mean Salary using numpy - it will be 4842684.1053811675 and store it in a variable called x
#MEAN
#to test this read in a fresh copy of your data file

# Calculate the mean of the "Salary" field
x = df['Salary'].mean()

print(x)

print("Mean: {:.2f}".format(x))

# Replace all NaN values in the "Salary" field with the mean salary
df['Salary'].fillna(x, inplace=True)
df


4842684.105381166
Mean: 4842684.11


Team  Number Position   Age Height  Weight        Salary
0    Boston Celtics     0.0       PG  25.0    6-2   180.0  7.730337e+06
1    Boston Celtics    99.0       SF  25.0    6-6   235.0  6.796117e+06
2    Boston Celtics    30.0       SG  27.0    6-5   205.0  4.842684e+06
3    Boston Celtics    28.0       SG  22.0    6-5   185.0  1.148640e+06
4    Boston Celtics     8.0       PF  29.0   6-10   231.0  5.000000e+06
..              ...     ...      ...   ...    ...     ...           ...
452       Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06
453       Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06
454       Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05
455       Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06
456       Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05

[457 rows x 7 columns]

In [27]:
#OR.... drop all occurences of rows with null values 
#using df.dropna() will drop null values in the whole dataset
#If necessary, do a google search to figure out how to drop all occurences of rows with null values in a specific column
dfSalary = df.dropna(subset=['Salary'])
dfSalary

#Re-run the ipynb file from the top (ignoring the cell above)
#Display the dataframe - row index 2 (the 3rd record is missing)


Team  Number Position   Age Height  Weight        Salary
0    Boston Celtics     0.0       PG  25.0    6-2   180.0  7.730337e+06
1    Boston Celtics    99.0       SF  25.0    6-6   235.0  6.796117e+06
2    Boston Celtics    30.0       SG  27.0    6-5   205.0  4.842684e+06
3    Boston Celtics    28.0       SG  22.0    6-5   185.0  1.148640e+06
4    Boston Celtics     8.0       PF  29.0   6-10   231.0  5.000000e+06
..              ...     ...      ...   ...    ...     ...           ...
452       Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06
453       Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06
454       Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05
455       Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06
456       Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05

[457 rows x 7 columns]

In [28]:
#Use count to show the number of values in all columns - it will help check that there are no null values
df.count()

Team        457
Number      457
Position    457
Age         457
Height      457
Weight      457
Salary      457
dtype: int64

In [29]:
#count & display null values in the Salary columns using isna() - there should be NO null values
nullCount = df['Salary'].isna().sum()
nullCount

0

In [30]:
#describe the dataframe
df.describe()

Number         Age      Weight        Salary
count  457.000000  457.000000  457.000000  4.570000e+02
mean    17.678337   26.938731  221.522976  4.842684e+06
std     15.966090    4.404016   26.368343  5.165781e+06
min      0.000000   19.000000  161.000000  3.088800e+04
25%      5.000000   24.000000  200.000000  1.100602e+06
50%     13.000000   26.000000  220.000000  2.869440e+06
75%     25.000000   30.000000  240.000000  6.331404e+06
max     99.000000   40.000000  307.000000  2.500000e+07

In [31]:
#Display the dataframe. In the Team column Boston Celtics appears in the first few records
df

Team  Number Position   Age Height  Weight        Salary
0    Boston Celtics     0.0       PG  25.0    6-2   180.0  7.730337e+06
1    Boston Celtics    99.0       SF  25.0    6-6   235.0  6.796117e+06
2    Boston Celtics    30.0       SG  27.0    6-5   205.0  4.842684e+06
3    Boston Celtics    28.0       SG  22.0    6-5   185.0  1.148640e+06
4    Boston Celtics     8.0       PF  29.0   6-10   231.0  5.000000e+06
..              ...     ...      ...   ...    ...     ...           ...
452       Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06
453       Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06
454       Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05
455       Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06
456       Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05

[457 rows x 7 columns]

In [32]:
#In the Team column change Boston Celtics to BC  
df.loc[df['Team'] == 'Boston Celtics', 'Team'] = 'BC'

#Display the dataframe
df

Team  Number Position   Age Height  Weight        Salary
0           BC     0.0       PG  25.0    6-2   180.0  7.730337e+06
1           BC    99.0       SF  25.0    6-6   235.0  6.796117e+06
2           BC    30.0       SG  27.0    6-5   205.0  4.842684e+06
3           BC    28.0       SG  22.0    6-5   185.0  1.148640e+06
4           BC     8.0       PF  29.0   6-10   231.0  5.000000e+06
..         ...     ...      ...   ...    ...     ...           ...
452  Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06
453  Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06
454  Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05
455  Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06
456  Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05

[457 rows x 7 columns]

In [33]:
#Filter the dataset to show all records for Team BC
df.loc[df['Team'].str.contains('BC')]

Team  Number Position   Age Height  Weight        Salary
0    BC     0.0       PG  25.0    6-2   180.0  7.730337e+06
1    BC    99.0       SF  25.0    6-6   235.0  6.796117e+06
2    BC    30.0       SG  27.0    6-5   205.0  4.842684e+06
3    BC    28.0       SG  22.0    6-5   185.0  1.148640e+06
4    BC     8.0       PF  29.0   6-10   231.0  5.000000e+06
5    BC    90.0       PF  29.0    6-9   240.0  1.200000e+07
6    BC    55.0       PF  21.0    6-8   235.0  1.170960e+06
7    BC    41.0        C  25.0    7-0   238.0  2.165160e+06
8    BC    12.0       PG  22.0    6-2   190.0  1.824360e+06
9    BC    36.0       PG  22.0    6-4   220.0  3.431040e+06
10   BC     7.0        C  24.0    6-9   260.0  2.569260e+06
11   BC     4.0       PG  27.0    5-9   185.0  6.912869e+06
12   BC    11.0       SG  27.0    6-7   220.0  3.425510e+06
13   BC    13.0       SG  20.0    6-6   215.0  1.749840e+06
14   BC    44.0        C  26.0    7-0   253.0  2.616975e+06

In [34]:
#Filter the dataset to show all records that don't contain Team BC
#This code will return all values in the Name column that DO NOT contain BC. 
df.loc[~df['Team'].str.contains('BC')]


Team  Number Position   Age Height  Weight     Salary
15   Brooklyn Nets    44.0       SG  27.0    6-8   216.0  3425510.0
16   Brooklyn Nets    22.0       SG  24.0    6-3   190.0   845059.0
17   Brooklyn Nets    21.0       SG  28.0    6-4   200.0  1500000.0
18   Brooklyn Nets    24.0       SG  21.0    6-7   220.0  1335480.0
19   Brooklyn Nets     2.0       PG  32.0    6-3   200.0  6300000.0
..             ...     ...      ...   ...    ...     ...        ...
452      Utah Jazz    41.0       PF  20.0   6-10   234.0  2239800.0
453      Utah Jazz     8.0       PG  26.0    6-3   203.0  2433333.0
454      Utah Jazz    25.0       PG  24.0    6-1   179.0   900000.0
455      Utah Jazz    21.0        C  26.0    7-3   256.0  2900000.0
456      Utah Jazz    24.0        C  26.0    7-0   231.0   947276.0

[442 rows x 7 columns]

In [35]:
#Count the number of individual values in the Position column - (you did somethimg similar earlier)
#You will see that there are 5 different positions
df["Position"].value_counts()


Position
SG    102
PF    100
PG     92
SF     85
C      78
Name: count, dtype: int64

In [36]:
#Convert the Position column to numeric values - we will use Get Dummies
  # when we use  the get_dummies function the data will be placed in 5 columns
#Use get_dummies on the Position column and assign the output to a temporary variable called dummies
dummies = pd.get_dummies(df['Position'], prefix = "dummies")

#Display the contents of dummies
dummies

dummies_C  dummies_PF  dummies_PG  dummies_SF  dummies_SG
0        False       False        True       False       False
1        False       False       False        True       False
2        False       False       False       False        True
3        False       False       False       False        True
4        False        True       False       False       False
..         ...         ...         ...         ...         ...
452      False        True       False       False       False
453      False       False        True       False       False
454      False       False        True       False       False
455       True       False       False       False       False
456       True       False       False       False       False

[457 rows x 5 columns]

In [37]:
#Add the temporary dummies columns to end of the existing dataframe df - concatenate - and update the dataframe
df = pd.concat([df, dummies], axis=1)

#Display the dataframe
df

Team  Number Position   Age Height  Weight        Salary  dummies_C  \
0           BC     0.0       PG  25.0    6-2   180.0  7.730337e+06      False   
1           BC    99.0       SF  25.0    6-6   235.0  6.796117e+06      False   
2           BC    30.0       SG  27.0    6-5   205.0  4.842684e+06      False   
3           BC    28.0       SG  22.0    6-5   185.0  1.148640e+06      False   
4           BC     8.0       PF  29.0   6-10   231.0  5.000000e+06      False   
..         ...     ...      ...   ...    ...     ...           ...        ...   
452  Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06      False   
453  Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06      False   
454  Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05      False   
455  Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06       True   
456  Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05       True   

     dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0         False        True       False       False  
1         False       False        True       False  
2         False       False       False        True  
3         False       False       False        True  
4          True       False       False       False  
..          ...         ...         ...         ...  
452        True       False       False       False  
453       False        True       False       False  
454       False        True       False       False  
455       False       False       False       False  
456       False       False       False       False  

[457 rows x 12 columns]

In [38]:
#At this point you might want to delete the Position column and update the dataframe. 
  #However when doing the visualisations later it might be handier to have it. 
  #We can always delete it before building the model
 
temp_df = df.drop(['Position'], axis=1)

#Display the dataframe

temp_df

Team  Number   Age Height  Weight        Salary  dummies_C  \
0           BC     0.0  25.0    6-2   180.0  7.730337e+06      False   
1           BC    99.0  25.0    6-6   235.0  6.796117e+06      False   
2           BC    30.0  27.0    6-5   205.0  4.842684e+06      False   
3           BC    28.0  22.0    6-5   185.0  1.148640e+06      False   
4           BC     8.0  29.0   6-10   231.0  5.000000e+06      False   
..         ...     ...   ...    ...     ...           ...        ...   
452  Utah Jazz    41.0  20.0   6-10   234.0  2.239800e+06      False   
453  Utah Jazz     8.0  26.0    6-3   203.0  2.433333e+06      False   
454  Utah Jazz    25.0  24.0    6-1   179.0  9.000000e+05      False   
455  Utah Jazz    21.0  26.0    7-3   256.0  2.900000e+06       True   
456  Utah Jazz    24.0  26.0    7-0   231.0  9.472760e+05       True   

     dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0         False        True       False       False  
1         False       False        True       False  
2         False       False       False        True  
3         False       False       False        True  
4          True       False       False       False  
..          ...         ...         ...         ...  
452        True       False       False       False  
453       False        True       False       False  
454       False        True       False       False  
455       False       False       False       False  
456       False       False       False       False  

[457 rows x 11 columns]

In [39]:
#Sort the dataframe by Age in ascending order and Weight in descending order and update the dataframe
temp_df = df.sort_values(['Age', 'Weight'], ascending=[True, False])
temp_df = df.drop(['Position'], axis=1)

#Display the dataframe
temp_df

Team  Number   Age Height  Weight        Salary  dummies_C  \
0           BC     0.0  25.0    6-2   180.0  7.730337e+06      False   
1           BC    99.0  25.0    6-6   235.0  6.796117e+06      False   
2           BC    30.0  27.0    6-5   205.0  4.842684e+06      False   
3           BC    28.0  22.0    6-5   185.0  1.148640e+06      False   
4           BC     8.0  29.0   6-10   231.0  5.000000e+06      False   
..         ...     ...   ...    ...     ...           ...        ...   
452  Utah Jazz    41.0  20.0   6-10   234.0  2.239800e+06      False   
453  Utah Jazz     8.0  26.0    6-3   203.0  2.433333e+06      False   
454  Utah Jazz    25.0  24.0    6-1   179.0  9.000000e+05      False   
455  Utah Jazz    21.0  26.0    7-3   256.0  2.900000e+06       True   
456  Utah Jazz    24.0  26.0    7-0   231.0  9.472760e+05       True   

     dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0         False        True       False       False  
1         False       False        True       False  
2         False       False       False        True  
3         False       False       False        True  
4          True       False       False       False  
..          ...         ...         ...         ...  
452        True       False       False       False  
453       False        True       False       False  
454       False        True       False       False  
455       False       False       False       False  
456       False       False       False       False  

[457 rows x 11 columns]

In [40]:
#Select rows where the age is greater than or equal to 20 and less than or equal to 29. Update the dataframe.
dfFiltered = df[(df['Age'] >= 20) & (df['Age'] <= 29)]

#Describe the dataframe
dfFiltered

Team  Number Position   Age Height  Weight        Salary  dummies_C  \
0           BC     0.0       PG  25.0    6-2   180.0  7.730337e+06      False   
1           BC    99.0       SF  25.0    6-6   235.0  6.796117e+06      False   
2           BC    30.0       SG  27.0    6-5   205.0  4.842684e+06      False   
3           BC    28.0       SG  22.0    6-5   185.0  1.148640e+06      False   
4           BC     8.0       PF  29.0   6-10   231.0  5.000000e+06      False   
..         ...     ...      ...   ...    ...     ...           ...        ...   
452  Utah Jazz    41.0       PF  20.0   6-10   234.0  2.239800e+06      False   
453  Utah Jazz     8.0       PG  26.0    6-3   203.0  2.433333e+06      False   
454  Utah Jazz    25.0       PG  24.0    6-1   179.0  9.000000e+05      False   
455  Utah Jazz    21.0        C  26.0    7-3   256.0  2.900000e+06       True   
456  Utah Jazz    24.0        C  26.0    7-0   231.0  9.472760e+05       True   

     dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0         False        True       False       False  
1         False       False        True       False  
2         False       False       False        True  
3         False       False       False        True  
4          True       False       False       False  
..          ...         ...         ...         ...  
452        True       False       False       False  
453       False        True       False       False  
454       False        True       False       False  
455       False       False       False       False  
456       False       False       False       False  

[333 rows x 12 columns]

In [41]:
#Filter the dataset to show players who earned more than 2 million, who were less than 25 and weighted more than 250

new_df = df[(df['Salary'] > 2000000) & (df['Age'] < 25) & (df['Weight'] > 250)]

new_df
#Assign results to new_df


#Display new_df


Team  Number Position   Age Height  Weight      Salary  \
10                      BC     7.0        C  24.0    6-9   260.0   2569260.0   
56      Philadelphia 76ers     8.0        C  20.0   6-11   275.0   4582680.0   
74         Toronto Raptors    17.0        C  24.0    7-0   255.0   4660482.0   
128           Phoenix Suns    21.0        C  22.0    7-1   260.0   3807120.0   
188        Detroit Pistons     0.0        C  22.0   6-11   279.0   3272091.0   
252        Houston Rockets     6.0       PF  24.0    6-9   252.0   2489530.0   
281   New Orleans Pelicans    23.0       PF  23.0   6-10   253.0   7070730.0   
412  Oklahoma City Thunder    12.0        C  22.0    7-0   255.0   2279040.0   
446              Utah Jazz    15.0       PF  24.0   6-10   265.0  12000000.0   

     dummies_C  dummies_PF  dummies_PG  dummies_SF  dummies_SG  
10        True       False       False       False       False  
56        True       False       False       False       False  
74        True       False       False       False       False  
128       True       False       False       False       False  
188       True       False       False       False       False  
252      False        True       False       False       False  
281      False        True       False       False       False  
412       True       False       False       False       False  
446      False        True       False       False       False

## Height is a Object type field - Get Height in Inches so that we can analyse it 

This section looks at stripping out the numeric values from the Height field (which is an object type)  
A calculation will be performed to work out total height in inches  
The results will be saved in a new numerical column and the original column will be deleted  

In [42]:
#Describe the object data in the df dataframe
df.describe(include = "O" )
#The only categorical/object type columns that we have is Height and Team


Team Position Height
count                    457      457    457
unique                    30        5     18
top     New Orleans Pelicans       SG    6-9
freq                      19      102     59

In [43]:
#Display the datatypes
temp_df.dtypes

Team           object
Number        float64
Age           float64
Height         object
Weight        float64
Salary        float64
dummies_C        bool
dummies_PF       bool
dummies_PG       bool
dummies_SF       bool
dummies_SG       bool
dtype: object

In [44]:
#View the first 5 lines to have a look at the Height column
temp_df.head()

Team  Number   Age Height  Weight        Salary  dummies_C  dummies_PF  \
0   BC     0.0  25.0    6-2   180.0  7.730337e+06      False       False   
1   BC    99.0  25.0    6-6   235.0  6.796117e+06      False       False   
2   BC    30.0  27.0    6-5   205.0  4.842684e+06      False       False   
3   BC    28.0  22.0    6-5   185.0  1.148640e+06      False       False   
4   BC     8.0  29.0   6-10   231.0  5.000000e+06      False        True   

   dummies_PG  dummies_SF  dummies_SG  
0        True       False       False  
1       False        True       False  
2       False       False        True  
3       False       False        True  
4       False       False       False

In [45]:
#To start we will split values from the Height columns and save the results in a new data frame called splitValues
#The values should be split on the hyphen"-"
    #splitValues[0] will contain the feet part of the strings
    #splitValues[1] will contain the inches part of the strings

#Split the height column and assign the results to a dataframe called splitValues
splitValues = df["Height"].str.split("-", n = 1, expand = True) 

#Display the dataframe called splitValues
splitValues

0   1
0    6   2
1    6   6
2    6   5
3    6   5
4    6  10
..  ..  ..
452  6  10
453  6   3
454  6   1
455  7   3
456  7   0

[457 rows x 2 columns]

In [46]:
#These values will be strings so convert them to numeric values

splitValues[0] = pd.to_numeric(splitValues[0])
splitValues[1] = pd.to_numeric(splitValues[1])

#calculate the total height in inches - (ft *12) + inches - and put the results in a new column called HeightInInches in the splitValues dataframe
splitValues['HeightInInches'] = splitValues[0] * 12 + splitValues[1]

#Display the dataframe called splitValues
splitValues


0   1  HeightInInches
0    6   2              74
1    6   6              78
2    6   5              77
3    6   5              77
4    6  10              82
..  ..  ..             ...
452  6  10              82
453  6   3              75
454  6   1              73
455  7   3              87
456  7   0              84

[457 rows x 3 columns]

In [47]:
#Concatenate the HeightInInches column from the splitValues dataframe with the df dataframe and save in df

#df = pd.concat([df, dummies], axis = 1)

#df = pd.concat([df, splitValues], axis = 1)
temp_df = pd.concat([df, splitValues['HeightInInches']], axis=1)

#Display the first 5 rows of df

temp_df.head()

Team  Number Position   Age Height  Weight        Salary  dummies_C  \
0   BC     0.0       PG  25.0    6-2   180.0  7.730337e+06      False   
1   BC    99.0       SF  25.0    6-6   235.0  6.796117e+06      False   
2   BC    30.0       SG  27.0    6-5   205.0  4.842684e+06      False   
3   BC    28.0       SG  22.0    6-5   185.0  1.148640e+06      False   
4   BC     8.0       PF  29.0   6-10   231.0  5.000000e+06      False   

   dummies_PF  dummies_PG  dummies_SF  dummies_SG  HeightInInches  
0       False        True       False       False              74  
1       False       False        True       False              78  
2       False       False       False        True              77  
3       False       False       False        True              77  
4        True       False       False       False              82

In [53]:
#In the dataframe df - drop original Height column 
#dfBroke = temp_df
#dfBroke.head()
# Remove the unwanted columns '0' and '1' from the dataframe
#f = df.drop(columns=[0, 1])
#temp_df.drop(columns=['Height'], inplace=True)

#Display the first 5 lines of df
temp_df.head()

Team  Number Position   Age  Weight        Salary  dummies_C  dummies_PF  \
0   BC     0.0       PG  25.0   180.0  7.730337e+06      False       False   
1   BC    99.0       SF  25.0   235.0  6.796117e+06      False       False   
2   BC    30.0       SG  27.0   205.0  4.842684e+06      False       False   
3   BC    28.0       SG  22.0   185.0  1.148640e+06      False       False   
4   BC     8.0       PF  29.0   231.0  5.000000e+06      False        True   

   dummies_PG  dummies_SF  dummies_SG  HeightInInches  
0        True       False       False              74  
1       False        True       False              78  
2       False       False        True              77  
3       False       False        True              77  
4       False       False       False              82

In [57]:
#Rearrange the dataframe so that HeightInInches appears after Salary and update the dataframe


#Display the first 5 lines of df. Where is the HeightInInches column now? 


# Assuming 'df' is your DataFrame
columns = list(temp_df.columns)

# Place the 'HeightInInches' right after the 'Salary' column
height_index = columns.index('HeightInInches')

columns.insert(columns.index('Salary') + 1, columns.pop(height_index))
temp_df = temp_df[columns]

# Display the first 5 lines of df
temp_df


Team  Number Position   Age  Weight        Salary  HeightInInches  \
0           BC     0.0       PG  25.0   180.0  7.730337e+06              74   
1           BC    99.0       SF  25.0   235.0  6.796117e+06              78   
2           BC    30.0       SG  27.0   205.0  4.842684e+06              77   
3           BC    28.0       SG  22.0   185.0  1.148640e+06              77   
4           BC     8.0       PF  29.0   231.0  5.000000e+06              82   
..         ...     ...      ...   ...     ...           ...             ...   
452  Utah Jazz    41.0       PF  20.0   234.0  2.239800e+06              82   
453  Utah Jazz     8.0       PG  26.0   203.0  2.433333e+06              75   
454  Utah Jazz    25.0       PG  24.0   179.0  9.000000e+05              73   
455  Utah Jazz    21.0        C  26.0   256.0  2.900000e+06              87   
456  Utah Jazz    24.0        C  26.0   231.0  9.472760e+05              84   

     dummies_C  dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0        False       False        True       False       False  
1        False       False       False        True       False  
2        False       False       False       False        True  
3        False       False       False       False        True  
4        False        True       False       False       False  
..         ...         ...         ...         ...         ...  
452      False        True       False       False       False  
453      False       False        True       False       False  
454      False       False        True       False       False  
455       True       False       False       False       False  
456       True       False       False       False       False  

[457 rows x 12 columns]

In [58]:
#Look at the Index references for each line in the output results
#Re-order the index values 
#Remember to use inplace=True 


#Display the dataframe

# Reset the index and drop the old one
temp_df.reset_index(drop=True, inplace=True)

# Display the dataframe
temp_df


Team  Number Position   Age  Weight        Salary  HeightInInches  \
0           BC     0.0       PG  25.0   180.0  7.730337e+06              74   
1           BC    99.0       SF  25.0   235.0  6.796117e+06              78   
2           BC    30.0       SG  27.0   205.0  4.842684e+06              77   
3           BC    28.0       SG  22.0   185.0  1.148640e+06              77   
4           BC     8.0       PF  29.0   231.0  5.000000e+06              82   
..         ...     ...      ...   ...     ...           ...             ...   
452  Utah Jazz    41.0       PF  20.0   234.0  2.239800e+06              82   
453  Utah Jazz     8.0       PG  26.0   203.0  2.433333e+06              75   
454  Utah Jazz    25.0       PG  24.0   179.0  9.000000e+05              73   
455  Utah Jazz    21.0        C  26.0   256.0  2.900000e+06              87   
456  Utah Jazz    24.0        C  26.0   231.0  9.472760e+05              84   

     dummies_C  dummies_PF  dummies_PG  dummies_SF  dummies_SG  
0        False       False        True       False       False  
1        False       False       False        True       False  
2        False       False       False       False        True  
3        False       False       False       False        True  
4        False        True       False       False       False  
..         ...         ...         ...         ...         ...  
452      False        True       False       False       False  
453      False       False        True       False       False  
454      False       False        True       False       False  
455       True       False       False       False       False  
456       True       False       False       False       False  

[457 rows x 12 columns]

In [59]:
#Filter the datset (df) to show the records for players taller than 85 inchches 
tall_players_df = temp_df[temp_df['HeightInInches'] > 85]

# Display the dataframe for tall players
tall_players_df


Team  Number Position   Age  Weight      Salary  \
40        New York Knicks     6.0       PF  20.0   240.0   4131720.0   
111    Los Angeles Lakers    17.0        C  29.0   270.0  15592217.0   
235      Dallas Mavericks    50.0        C  29.0   245.0    525093.0   
275  New Orleans Pelicans    42.0        C  28.0   248.0   4389607.0   
302     San Antonio Spurs    40.0        C  27.0   290.0   1200000.0   
322         Atlanta Hawks    22.0        C  24.0   260.0   1000000.0   
455             Utah Jazz    21.0        C  26.0   256.0   2900000.0   

     HeightInInches  dummies_C  dummies_PF  dummies_PG  dummies_SF  dummies_SG  
40               87      False        True       False       False       False  
111              86       True       False       False       False       False  
235              86       True       False       False       False       False  
275              86       True       False       False       False       False  
302              87       True       False       False       False       False  
322              87       True       False       False       False       False  
455              87       True       False       False       False       False

In [60]:
#Save the dataframe as a new dataset.
#Index = false is used to ensure that the row index values are not included in the dataset
# Save df to a new CSV file without the index
temp_df.to_csv('new_dataset.csv', index=False)
